server/src/main/java/org/elasticsearch/index/mapper/ContentPath.java

```diff
@@ -27,18 +31,37 @@ public ContentPath() {
     public void add(String name) {
         path[index++] = name;
         if (index == path.length) { // expand if needed
-            expand();
+            expandPath();
+        }
+    }
+
+    public void addDottedFieldName(String name) {
+        dottedFieldName[dottedFieldNameIndex++] = name;
+        if (dottedFieldNameIndex == dottedFieldName.length) {
+            expandDottedFieldName();
         }
     }
 
-    private void expand() {
+    private void expandPath() {
         String[] newPath = new String[path.length + 10];
         System.arraycopy(path, 0, newPath, 0, path.length);
         path = newPath;
     }
 
-    public void remove() {
-        path[index--] = null;
```

Comment:
This was a wrong behaviour, index has the next available index to write path to. In fact to add a path we did `path[index++] = name;` so we write at index position and increment it later. Doing `path[index--] = null;` would set to null the next available index position (that is already null) and later subtract 1 from index. This was masked by the fact that when retrieving a path we loop up until `i < index`, therefore masking the fact that there could have being 'removed' but not null path existing at `index`.

Reply:
could you extract this fix out of this PR and add a test for it? I am not clear on whether it causes actual issues, and if the fix is needed to make your changes work. Still, I think it would be worthwhile to pull it out of this PR.

Reply:
is that a pre-requisite for this change or a separate one? Can this PR go in without it?